crates/cairo-coverage/src/input/filter.rs

```diff
@@ -0,0 +1,90 @@
+use crate::cli::IncludedComponent;
+use crate::data_loader::LoadedData;
+use crate::input::sierra_to_cairo_map::StatementOrigin;
+use regex::Regex;
+use std::collections::HashSet;
+use std::iter::once;
+use std::sync::LazyLock;
+
+pub static VIRTUAL_FILE_REGEX: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"\[.*?]").unwrap());
+const SNFORGE_TEST_EXECUTABLE: &str = "snforge_internal_test_executable";
+
+#[derive(Eq, PartialEq, Hash)]
+enum ItemLabel {
+    TestFunction,
+    UserFunction,
+    NonUserFunction,
+    Macro,
+}
+
+impl From<IncludedComponent> for ItemLabel {
+    fn from(included_component: IncludedComponent) -> Self {
+        match included_component {
+            IncludedComponent::TestFunctions => ItemLabel::TestFunction,
+            IncludedComponent::Macros => ItemLabel::Macro,
+        }
+    }
+}
+
+pub struct Filter {
+    allowed_labels: HashSet<ItemLabel>,
+    test_functions: HashSet<String>,
+}
+
+impl Filter {
+    pub fn new(included_component: &[IncludedComponent], loaded_data: &LoadedData) -> Self {
+        let test_functions = loaded_data
+            .debug_info
+            .executables
+            .get(SNFORGE_TEST_EXECUTABLE)
+            .unwrap_or(&Vec::new())
+            .iter()
+            .map(ToString::to_string)
+            .collect();
+
+        let allowed_labels = included_component
+            .iter()
+            .cloned()
+            .map(ItemLabel::from)
+            .chain(once(ItemLabel::UserFunction))
+            .collect();
+
+        Self {
+            allowed_labels,
+            test_functions,
+        }
+    }
+
+    pub fn should_include(&self, statement_origin: &StatementOrigin) -> bool {
+        self.labels(statement_origin)
+            .is_subset(&self.allowed_labels)
+    }
+
+    fn labels(
+        &self,
+        StatementOrigin {
+            function_name,
+            file_location,
+            ..
+        }: &StatementOrigin,
+    ) -> HashSet<ItemLabel> {
+        let mut labels = HashSet::new();
+
+        if self.test_functions.contains(function_name) {
+            labels.insert(ItemLabel::TestFunction);
+        } else if VIRTUAL_FILE_REGEX.is_match(file_location) {
```

Comment:
Could you explain this logic?

Reply:
> The key difference is that macros and similar constructs include additional information in square brackets `[]` as  they are virtual files.
>
> #### Examples
>
>**Normal function path:**
>```
>/path/to/user/project/src/lib.cairo
>```
>
>**Virtual files:**
>```
>/path/to/user/project/src/lib.cairo[some_info]
>```
>
>For a write macro, the path would be:
>```
>/path/to/user/project/src/lib.cairo[write_macro]
>```

Description from this https://github.com/software-mansion/cairo-coverage/issues/34.

So if something is macro it will be matched against this regex. Not ideal but the best we can do with current mappings. There are plans to change it to more descriptive mappings https://github.com/software-mansion/cairo-coverage/issues/44

Reply:
> macros and similar constructs


What are "similar constructs"?

Reply:
The only other example is tests cases in compiler repository, there are put in virtual files. But i doubt that compiler team will use coverage. Maybe i should write just macros there after i think about this 🧐

Reply:
Ok thanks for explanation :)